src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class ServiceScaleManager
+    {
+        private readonly ILogger _logger;
+        private bool _inited = false;
+
+        private IReadOnlyList<ServiceEndpoint> _endpointsStore = new List<ServiceEndpoint>();
+
+        public ServiceScaleManager(ILoggerFactory loggerFactory,
+            IOptionsMonitor<ServiceOptions> optionsMonitor)
+        {
+            _logger = loggerFactory?.CreateLogger<ServiceScaleManager>();
+            OnChange(optionsMonitor.CurrentValue);
+            optionsMonitor.OnChange(OnChange);
+
+            _endpointsStore = optionsMonitor.CurrentValue.Endpoints;
+            _inited = true;
+        }
+
+        private void OnChange(ServiceOptions options)
+        {
+            // Skip init app starts and respect EnableAutoScale flag
+            if (options.EnableAutoScale && _inited)
+            {
+                Log.DetectEndpointChanges(_logger);
+
+                var updatedEndpoints = options.Endpoints;
+                var endpoints = GetChangedEndpoints(updatedEndpoints);
+
+                // Add then remove to minor affect to new clients
+                OnAdd(endpoints.AddedEndpoints);
+
+                OnRemove(endpoints.RemovedEndpoints);
```

Comment:
>OnRemove [](start = 16, length = 8)

`add` and `remove` can be done parallelly?

Reply:
I think better to wait add complete cause add need to wait server match before adding negotiation and remove will remove negotiation directly, which means before adding successfully, there'll be few endpoints available.

---
In reply to: [368272199](https://github.com/Azure/azure-signalr/pull/801#discussion_r368272199) [](ancestors = 368272199)